File: src/ai/ai_player.ts
import { Card, Rank, Suit } from '../models/card';
import { Hand, handSize, countByRank, countBySuit } from '../models/hand';
import { GameState, PlayerAction } from '../models/game_state';
import { analyzeHand, wouldFormChan, wouldFormCa, canChiu } from '../engine/meld_analyzer';
import { GameEngine } from '../engine/game_engine';

export interface AiDecision {
  action: PlayerAction;
  cardId?: string;
}

export interface AiPlayer {
  decide(state: GameState, playerId: number, engine: GameEngine): AiDecision;
}

function cardValue(card: Card, hand: Hand, state: GameState): number {
  let value = 0;

  const sameRankCount = countByRank(hand, card.rank);
  value += sameRankCount * 10;

  const sameSuitCount = countBySuit(hand, card.suit);
  value += sameSuitCount * 3;

  const discardedCount = state.deck.discardPile.filter(
    c => c.rank === card.rank && c.suit === card.suit
  ).length;
  // If 3 copies are already discarded, this card is dead weight (can never form Chăn)
  if (discardedCount >= 3) {
    value -= 50;
  } else if (discardedCount === 2) {
    value -= 15;
  }

  // Bonus for scoring cards
  if (card.rank === Rank.Chi) value += 8;
  if (card.rank === Rank.Cuu && card.suit === Suit.Van) value += 3;
  if (card.rank === Rank.That && card.suit === Suit.Van2) value += 3;

  // Penalty for isolated cards (no same-rank, no same-suit neighbor)
  const hasSameRank = sameRankCount > 1;
  const hasSameSuit = hand.cards.some(
    c => c.suit === card.suit && c.rank !== card.rank
  );
  if (!hasSameRank && !hasSameSuit) {
    value -= 15;
  }

  return value;
}

function isSafeDiscard(card: Card, state: GameState, playerId: number): boolean {
  const discardCount = state.deck.discardPile.filter(
    c => c.rank === card.rank && c.suit === card.suit
  ).length;

  // If 3+ of this card type are visible, it's safe
  if (discardCount >= 3) return true;

  // Chi Chi is risky
  if (card.rank === Rank.Chi) return false;

  // Edge ranks (2, 9) are slightly safer
  if (card.rank === Rank.Nhi || card.rank === Rank.Cuu) return true;

  // Cards that have been discarded once are somewhat safe
  if (discardCount === 1) return true;

  return false;
}

// Evaluate how many "steps" a card is from becoming useful
function discardDangerScore(
  card: Card,
  hand: Hand,
  state: GameState,
  analysis: ReturnType<typeof analyzeHand>,
  playerId: number
): number {
  let danger = 0;

  // Is the card part of a Chăn? (highest priority to keep)
  const inChan = analysis.chans.some(ch => ch.cards.some(c => c.id === card.id));
  if (inChan) return 1000; // Never discard Chăn

  // Is the card part of a Cạ?
  const inCa = analysis.cas.some(ca => ca.cards.some(c => c.id === card.id));
  if (inCa) danger += 40;

  // Is the card part of a Ba Đầu?
  const inBaDau = analysis.baDaus.some(bd => bd.cards.some(c => c.id === card.id));
  if (inBaDau) danger += 30;

  // How many cards of same rank? (2 = potential Chăn, 3 = can Chiu)
  const sameRank = countByRank(hand, card.rank);
  danger += sameRank * 15;

  // How many cards of same suit? (potential Cạ)
  const sameSuit = countBySuit(hand, card.suit);
  danger += sameSuit * 5;

  // Safe cards are less dangerous to discard
  if (isSafeDiscard(card, state, playerId)) {
    danger -= 20;
  }

  // Already discarded copies make it safer
  const discardedCount = state.deck.discardPile.filter(
    c => c.rank === card.rank && c.suit === card.suit
  ).length;
  danger -= discardedCount * 10;

  return danger;
}

function chooseDiscard(hand: Hand, state: GameState, playerId: number): Card | null {
  const analysis = analyzeHand(hand);

  const chanCardIds = new Set(
    analysis.chans.flatMap(ch => ch.cards.map(c => c.id))
  );
  const discardable = hand.cards.filter(c => !chanCardIds.has(c.id));

  if (discardable.length === 0) return null;

  const scored = discardable.map(card => ({
    card,
    score: evaluateDiscardScore(card, hand, state, analysis, playerId),
  }));

  scored.sort((a, b) => a.score - b.score);
  return scored[0].card;
}

function evaluateDiscardScore(
  card: Card,
  hand: Hand,
  state: GameState,
  analysis: ReturnType<typeof analyzeHand>,
  playerId: number
): number {
  let score = 0;

  // Value of keeping the card
  score -= cardValue(card, hand, state);

  // Danger of discarding (lower = safer to discard)
  score += discardDangerScore(card, hand, state, analysis, playerId);

  // Bonus for safe discards
  if (isSafeDiscard(card, state, playerId)) {
    score -= 15;
  }

  return score;
}

export function createAiPlayer(): AiPlayer {
  return {
    decide(state: GameState, playerId: number, engine: GameEngine): AiDecision {
      const player = state.players[playerId];
      const hand = player.hand;
      const actions = engine.getValidActions(state, playerId);

      if (actions.includes(PlayerAction.DeclareU)) {
        return { action: PlayerAction.DeclareU };
      }

      if (actions.includes(PlayerAction.Chiu)) {
        return { action: PlayerAction.Chiu };
      }

      if (actions.includes(PlayerAction.Eat)) {
        const lastDiscard = state.deck.discardPile[state.deck.discardPile.length - 1];
        if (lastDiscard) {
          const eatOpts = engine.getEatOptions(state, playerId);
          if (eatOpts.length > 0) {
            if (wouldFormChan(hand, lastDiscard)) {
              return { action: PlayerAction.Eat };
            }
            if (wouldFormCa(hand, lastDiscard)) {
              const analysis = analyzeHand(hand);
              if (analysis.chanCount < 6) {
                return { action: PlayerAction.Eat };
              }
            }
          }
        }
      }

      if (actions.includes(PlayerAction.Draw)) {
        return { action: PlayerAction.Draw };
      }

      if (actions.includes(PlayerAction.Discard)) {
        const discardCard = chooseDiscard(hand, state, playerId);
        if (discardCard) {
          return { action: PlayerAction.Discard, cardId: discardCard.id };
        }
      }

      return { action: PlayerAction.Pass };
    },
  };
}

File: src/ChanDanGian.Core/AI/AiPlayer.cs
namespace ChanDanGian.Core.AI;

using ChanDanGian.Core.Engine;
using ChanDanGian.Core.Models;

public record AiDecision(PlayerAction Action, string? CardId = null);

public interface IAiPlayer
{
    AiDecision Decide(GameState state, int playerId, IGameEngine engine);
}

public class AiPlayer : IAiPlayer
{
    public AiDecision Decide(GameState state, int playerId, IGameEngine engine)
    {
        var player = state.Players[playerId];
        var hand = player.Hand;
        var actions = engine.GetValidActions(state, playerId);

        // Priority 1: Declare U
        if (actions.Contains(PlayerAction.DeclareU))
            return new AiDecision(PlayerAction.DeclareU);

        // Priority 2: Chiu
        if (actions.Contains(PlayerAction.Chiu))
            return new AiDecision(PlayerAction.Chiu);

        // Priority 3: Eat (if forms Chan or Ca and we have room)
        if (actions.Contains(PlayerAction.Eat))
        {
            var lastDiscard = state.Deck.TopDiscard;
            if (lastDiscard != null)
            {
                var eatOptions = engine.GetEatOptions(state, playerId);
                if (eatOptions.Count > 0)
                {
                    if (MeldAnalyzer.WouldFormChan(hand, lastDiscard))
                        return new AiDecision(PlayerAction.Eat);

                    if (MeldAnalyzer.WouldFormCa(hand, lastDiscard))
                    {
                        var analysis = MeldAnalyzer.AnalyzeHand(hand);
                        if (analysis.ChanCount < 6)
                            return new AiDecision(PlayerAction.Eat);
                    }
                }
            }
        }

        // Priority 4: Draw
        if (actions.Contains(PlayerAction.Draw))
            return new AiDecision(PlayerAction.Draw);

        // Priority 5: Discard (choose lowest danger card)
        if (actions.Contains(PlayerAction.Discard))
        {
            var discardCard = ChooseDiscard(hand, state, playerId);
            if (discardCard != null)
                return new AiDecision(PlayerAction.Discard, discardCard.Id);
        }

        // Priority 6: Pass
        return new AiDecision(PlayerAction.Pass);
    }

    private static Card? ChooseDiscard(Hand hand, GameState state, int playerId)
    {
        var analysis = MeldAnalyzer.AnalyzeHand(hand);
        var chanCardIds = analysis.Chans
            .SelectMany(ch => ch.Cards.Select(c => c.Id))
            .ToHashSet();
        var discardable = hand.Cards.Where(c => !chanCardIds.Contains(c.Id)).ToList();

        if (discardable.Count == 0) return null;

        var scored = discardable
            .Select(c => (Card: c, Score: EvaluateDiscardScore(c, hand, state, analysis, playerId)))
            .ToList();

        return scored.OrderBy(s => s.Score).First().Card;
    }

    private static int CardValue(Card card, Hand hand, GameState state)
    {
        int value = 0;

        var sameRankCount = hand.CountByRank(card.Rank);
        value += sameRankCount * 10;

        var sameSuitCount = hand.CountBySuit(card.Suit);
        value += sameSuitCount * 3;

        var discardedCount = state.Deck.DiscardPile
            .Count(c => c.Rank == card.Rank && c.Suit == card.Suit);

        if (discardedCount >= 3)
            value -= 50;
        else if (discardedCount >= 2)
            value -= 15;

        // Scoring card bonus
        if (card.Rank == Rank.Chi) value += 8;
        if (card.Rank == Rank.Cuu && card.Suit == Suit.Van) value += 3;
        if (card.Rank == Rank.That && card.Suit == Suit.Van2) value += 3;

        // Isolation penalty
        bool hasSameRank = sameRankCount > 1;
        bool hasSameSuit = hand.Cards.Any(c => c.Suit == card.Suit && c.Rank != card.Rank);
        if (!hasSameRank && !hasSameSuit)
            value -= 15;

        return value;
    }

    private static bool IsSafeDiscard(Card card, GameState state, int playerId)
    {
        var discardCount = state.Deck.DiscardPile
            .Count(c => c.Rank == card.Rank && c.Suit == card.Suit);

        if (discardCount >= 3) return true;
        if (card.Rank == Rank.Chi) return false;
        if (card.Rank == Rank.Nhi || card.Rank == Rank.Cuu) return true;
        if (discardCount >= 1) return true;

        return false;
    }

    private static int EvaluateDiscardScore(
        Card card, Hand hand, GameState state,
        HandAnalysis analysis, int playerId)
    {
        int score = 0;

        // Value of keeping (negative = valuable)
        score -= CardValue(card, hand, state);

        // Danger of discarding
        score += DiscardDangerScore(card, hand, state, analysis, playerId);

        // Safe discard bonus
        if (IsSafeDiscard(card, state, playerId))
            score -= 15;

        return score;
    }

    private static int DiscardDangerScore(
        Card card, Hand hand, GameState state,
        HandAnalysis analysis, int playerId)
    {
        int danger = 0;

        // In Chan? Never discard
        if (analysis.Chans.Any(ch => ch.Cards.Any(c => c.Id == card.Id)))
            return 1000;

        // In Ca?
        if (analysis.Cas.Any(ca => ca.Cards.Any(c => c.Id == card.Id)))
            danger += 40;

        // In Ba Dau?
        if (analysis.BaDaus.Any(bd => bd.Cards.Any(c => c.Id == card.Id)))
            danger += 30;

        // Same rank count
        var sameRank = hand.CountByRank(card.Rank);
        danger += sameRank * 15;

        // Same suit count
        var sameSuit = hand.CountBySuit(card.Suit);
        danger += sameSuit * 5;

        // Safe card
        if (IsSafeDiscard(card, state, playerId))
            danger -= 20;

        // Already discarded copies
        var discardedCount = state.Deck.DiscardPile
            .Count(c => c.Rank == card.Rank && c.Suit == card.Suit);
        danger -= discardedCount * 10;

        return danger;
    }
}
